### PR TITLE
Implement freshWord cheatcode

### PR DESCRIPTION
**Pull request by iFrostizz**
Closes: https://github.com/runtimeverification/evm-semantics/issues/1905

Adds a new `freshWord` kevm cheatcode for returning a symbolic 32 bytes word in a foundry test.
~~Support for other types is then done by bit masking at the evm level.~~
We will support narrower types by writing the expected bytes width as the argument.
Signed int support is done by explicitly declaring a new cheatcode.

### CODE REVIEW

File: include/kframework/foundry.md
Comment (ehildenb) [inline include/kframework/foundry.md:570]:
```suggestion
```k
    rule [foundry.call.freshWord]:
         <k> #call_foundry SELECTOR _ => . ... </k>
         <output> _ => #bufStrict(32, WORD) </output>
      requires SELECTOR ==Int selector ( "freshWord()" )
       andBool 0 <=Int WORD
     ensures #rangeUInt(256, WORD)
```
```

**Comment by iFrostizz** [comment]
Added a new `freshBool` cheatcode because it had less overhead, but maybe that we would only create a `kevm.freshBool()` that returns `true` if the output is `> 0 ` and `false` otherwise ?

File: include/kframework/foundry.md
Comment (lucasmt) [inline include/kframework/foundry.md:1474]:
```suggestion
    rule ( selector ( "freshSInt(uint8)" )                         => 978336391  )
```
Probably a copy-paste mistake, `625253732` is the selector for `freshUInt(uint8)` below

File: include/kframework/foundry.md
Comment (lucasmt) [inline include/kframework/foundry.md:567]: This doesn't work because `#bufStrict(SIZE, DATA)` requires `0 <=Int DATA`. I don't know if it makes sense to have a signed version of this cheatcode, instead it might be better to just implement `freshSInt...()` like
```
function freshSInt256() external returns (int256) {
    return int256(kevm.freshUInt256());
}
```

Comment (iFrostizz) [reply include/kframework/foundry.md:567]: Right, nice catch
I don't know if there is less overhead with a native cheatcode ? It seems like there isn't any, we might defer to @ehildenb

Comment (ehildenb) [reply include/kframework/foundry.md:567]: Native cheatcodes will usually have lower overhead, because they don't require executing any code aside from calling into the KEVMCheats contract.

Either approach seems fine.

File: kevm-pyk/poetry.lock
Comment (anvacaru) [inline kevm-pyk/poetry.lock:1]: Is there a reason to downgrade the Poetry version and modify this file?

Comment (iFrostizz) [reply kevm-pyk/poetry.lock:1]: This was not intentional, will back out

Comment (iFrostizz) [reply kevm-pyk/poetry.lock]: thank you !

File: include/kframework/foundry.md
Comment (anvacaru) [inline include/kframework/foundry.md:197]: Declaring the `FOUNDRY-CHEAT-CODES` module `symbolic` will block the booster compilation. Even if the `symbolic` tag is removed, the compilation is still blocked by existential vars.

File: tests/foundry/test/FreshInt.t.sol
Comment (ehildenb) [inline tests/foundry/test/FreshInt.t.sol:31]: Can we add a failing test, which asserts something like `assertLe(val, max + 1)`, and then actually check the output of it, by adding it to the `tests/foundry/checkoutput` list?

File: include/kframework/foundry.md
Comment (ehildenb) [inline include/kframework/foundry.md:587]:
```suggestion
    rule [foundry.call.freshBool]:
         <k> #call_foundry SELECTOR _ => . ... </k>
         <output> _ => #bufStrict(32, ?WORD) </output>
      requires SELECTOR ==Int selector ( "freshBool()" )
       ensures #rangeBool(?WORD)
```

File: include/kframework/lemmas/bytes-simplification.k
Comment (ehildenb) [inline include/kframework/lemmas/bytes-simplification.k:30]: What number is this? We should give it a name:

image

It's 32 `F` followed by one `8` followed by 31 `0` in hex.

Comment (anvacaru) [reply include/kframework/lemmas/bytes-simplification.k:30]: @iFrostizz discovered that it is `maxUInt256 - maxUInt128 / 2`.

Comment (iFrostizz) [reply include/kframework/lemmas/bytes-simplification.k:30]: This is also `not` `type(int128).max`
image
So it's `type(int128).min`
image

Comment (iFrostizz) [reply include/kframework/lemmas/bytes-simplification.k:30]: In the signed form, it's `typeint(128).min`
image

File: include/kframework/lemmas/bytes-simplification.k
Comment (anvacaru) [inline include/kframework/lemmas/bytes-simplification.k:33]:
```suggestion
    rule [signextend-min]: (signextend( _ , W:Int) s<Word minSInt128Word ) => 0 [simplification, symbolic(W)]
    rule [signextend-max]: (maxSInt128 s<Word signextend( _ , W:Int) ) => 0 [simplification, symbolic(W)]
```

**Review by anvacaru — APPROVED** [review]

**Pull request merged**
